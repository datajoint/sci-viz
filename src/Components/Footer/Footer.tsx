import { useState, useEffect } from 'react'
import { version } from '../../../package.json'

import './Footer.css'

function Footer() {
    const [backendVersion, setBackendVersion] = useState('')

    async function getVersionNumber() {
        let basePath = window.location.href.split('/')
        basePath.pop()
        const response = await fetch(
            `${basePath.join('/')}${process.env.REACT_APP_DJSCIVIZ_BACKEND_PREFIX}/version`,
            {
                method: 'GET',
                headers: { 'Content-Type': 'application/json' }
            }
        )

        if (response.status === 500) {
            throw new Error('Unable to get version number')
        }

        return response.json()
    }

    useEffect(() => {
        getVersionNumber()
            .then((result) => {
                if (result.version) {
                    setBackendVersion(result.version)
                }
            })
            .catch((error) => {
                setBackendVersion('Unable to get version number')
            })
    }, [])

    return (
        <footer>
            <div className='footer-content'>
                <p>Â© 2023, DataJoint SciViz</p>
            </div>
            <div className='version-info-div'>
                <div className='version-number'>
                    <b>Front End Version:</b> {version} <b>Back End Version:</b>{' '}
                    {backendVersion}
                </div>
            </div>
        </footer>
    )
}

export default Footer
