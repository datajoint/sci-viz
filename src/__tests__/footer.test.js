import renderer from 'react-test-renderer'
import { render } from '@testing-library/react'
import '@testing-library/jest-dom'
import packageJSON from '../../package.json'

import Footer from '../Components/Footer/Footer.tsx'
global.fetch = jest.fn(() => {})

afterEach(() => {
    jest.restoreAllMocks()
})

// Snapshot
it('snapshot_of_footer_component', () => {
    const tree = renderer.create(<Footer />).toJSON()
    expect(tree).toMatchSnapshot()
})

// Tests that the backend version number is set successfully.
it('test_set_backend_version_successfully', async () => {
    const mockResponse = { version: packageJSON.version }
    jest.spyOn(global, 'fetch').mockResolvedValue({
        json: jest.fn().mockResolvedValue(mockResponse),
        status: 200
    })

    const { findByText } = render(<Footer />)
    const versionNumber = await findByText(packageJSON.version)

    expect(versionNumber).toBeInTheDocument()

    global.fetch.mockRestore()
})

// Tests that the function fails to fetch the version number from the backend but gracefully handles the error and sets an appropriate message in the state.
it('test_get_version_number_failure', async () => {
    jest.spyOn(global, 'fetch').mockResolvedValueOnce({
        status: 500
    })

    const { findByText } = render(<Footer />)
    const versionNumber = await findByText('Unable to get version number', { exact: false })

    expect(versionNumber).toBeInTheDocument()
})
