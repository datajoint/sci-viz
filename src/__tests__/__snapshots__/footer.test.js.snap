// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot_of_footer_component 1`] = `
<footer>
  <div
    className="footer-content"
  >
    <p>
      Â© 2023, DataJoint SciViz
    </p>
  </div>
  <div
    className="version-info-div"
  >
    <div
      className="version-number"
    >
      <b>
        Front End Version:
      </b>
       
      2.3.4
       
      <b>
        Back End Version:
      </b>
       
    </div>
  </div>
</footer>
`;
